AI for basic mobs, slightly harder mobs, random optimizations, partial action functionality, mobs deal with corpses

crix870 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->

## About The Pull Request

Ports basic mobs from Vanderlin as well as some AI components from TG/Monkeystation.

https://github.com/Vanderlin-Tales-Of-Wine/Vanderlin/pull/48

The original intent behind porting this system was to help me familiarize myself with, learn and understand the intricacies of SS13’s code—and it has successfully served that purpose. I'll probably tinker around with it here and there for my own fun.

I encourage you if you have some time and motivation to play around with it, test, debug etc. I've done one or two runs of the wilds/orc dungeon and didn't have any major (non-skill) issues.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Honestly I'm not sure if it is worth it. The current system in Azure Peak is functional with no major performance issues and it's unclear (to me) whether the potential performance improvements (?) justify the inevitable AI bugs that will arise. Still, it's here if anyone's interested.

Key problems at the moment:
~~- Mobs when running away won't attack back. Compared to the original AI, mobs at some point will turn around and attack you. But in this version, they will keep running until the end of the map, a barrier or if they fall a Z-level.~~

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

---- NEW

Complete list of changes.
-   New AI system for basic/simple mobs - the infrastructure is what's important.
-   Hostile mobs (not farm animals) are ported over to this new system.
-   Action system partially re-implemented for 2 mobs. It needs to be heavily fleshed out but is in working state for AI
-   Dragons have additional or changed functionality. (dragonbreath, dragon leap)
-   Additional turf helpers (get_ranged_target_turf_direct and slightly faster than the original mob checks for traversability/pathfinding (turf.can_traverse).
-  Slightly optimized can_see function, will check distance first, saved a single byte (maybe?) on the loop.
-  click function now allows for non-carbons to have weapon reach (this is stored in the intents).
-  Some mobs will now eat (butcher) corpses. (wolves, trolls, rats, spiders, dragons, minotaurs, moles)
-  Some additional mobs will now eat food. (rats, wolves, trolls)
   Currently this does nothing for the mob except sate their hunger which is also currently used for nothing (as they're non-farm animals). Could probably work in spawning more or healing the mob later on.
- Added functionality to some mobs to sit and heal themselves when they have no target. (staunch bleeding, heal self). Current amounts are 0.01% for bleed, 0.1% for brute. Note, this does not refill blood.
- Added a little bit of functionality to reduce flammability of non-flammable simple mobs (dragon).
- Cleaned up a little of the retaliate mob code.
- There was an old minotaur code file, don't know it's purpose but it's commented out for now. ([code/modules/mob/living/simple_animal/rogue/creacher/minotaur.dm](https://github.com/GeneralPantsuIsBadAtCoding/Azure-Peak/pull/996/files#diff-b38b58b374339dc1dfb3142eaaec05c3e329bbb50692bea8906a0d9a91f3e594))
- Reduced the chances of mobs idle sound playing slightly
- Altered the damage balance of the penultimate mobs slightly (minotaur, dragon) - this may need to be tweaked up further. Increased the reach of their weapons from 1 to 2
- Troll punches increased slightly.
- Trialling improved AI marauder (spearmen) orcs. I think this code is somewhat taxing.
- Dragon firebreath shifted over to spell but is unused at the moment. (chose action system as spells seemed to be a little bit of a mess. Can easily change over in the future if resolved)
- Some runtime errors fixed with carbon mobs
- Fixed some typos, magic numbers in code where I encountered them

Known Issues
- If you pause the game while debugging it, it may spew out a multitude of null errors from the AI system - a timer or something. I'm unsure as of yet what causes it as the message is not helpful.
- Shooting a mob from very far away, 1+ screens will generally not 'aggro' or activate it. 
- I don't know best practices for BYOND code so some may code not be as efficient where it can be. 
- Multiplayer performance: unknown, single player: within acceptable limits

bearrrrrrrr commented:
HYPE

dwasint commented:
their is an easy fix for them running away, and thats just adding a new var to the flee element that makes them stop fleeing after a certain point

dwasint commented:
but the real reason to port this is it allows better control over what mobs do as you can give them more indepth plans that aren't a pain to read

crix870 commented:
> but the real reason to port this is it allows better control over what mobs do as you can give them more indepth plans that aren't a pain to read

Yup, I've realised this over the past few days -- especially with small tweaks for differing mob subtypes when needed. And it cuts down significantly on repetitive code.
 
Thanks for your work/inspo!

dwasint commented:
> > but the real reason to port this is it allows better control over what mobs do as you can give them more indepth plans that aren't a pain to read
> 
> Yup, I've realised this over the past few days -- especially with small tweaks for differing mob subtypes when needed. And it cuts down significantly on repetitive code.
> 
> Thanks for your work/inspo!

if you get stuck I did make a pr on vanderlin a few hours ago redoing some of this code to help with the issue

crix870 commented:
I'm mostly satisfied with the features implemented and I learned a lot about BYOND. There -should- be very little conflicts with the debloat but I can't say with certainty. List of changes/issues thus far.

-   New AI system for basic/simple mobs - the infrastructure is what's important.
-   Hostile mobs (not farm animals) are ported over to this new system.
-   Action system partially re-implemented for 2 mobs. It needs to be heavily fleshed out but is in working state for AI
-   Dragons have additional or changed functionality. (dragonbreath, dragon leap)
-   Additional turf helpers (get_ranged_target_turf_direct and slightly faster than the original mob checks for traversability/pathfinding (turf.can_traverse).
-  Slightly optimized can_see function, will check distance first, saved a single byte (maybe?) on the loop.
-  click function now allows for non-carbons to have weapon reach (this is stored in the intents).
-  Some mobs will now eat (butcher) corpses. (wolves, trolls, rats, spiders, dragons, minotaurs, moles)
-  Some additional mobs will now eat food. (rats, wolves, trolls)
   Currently this does nothing for the mob except sate their hunger which is also currently used for nothing (as they're non-farm animals). Could probably work in spawning more or healing the mob later on.
- Added functionality to some mobs to sit and heal themselves when they have no target. (staunch bleeding, heal self). Current amounts are 0.01% for bleed, 0.1% for brute. Note, this does not refill blood.
- Added a little bit of functionality to reduce flammability of non-flammable simple mobs (dragon).
- Cleaned up a little of the retaliate mob code.
- There was an old minotaur code file, don't know it's purpose but it's commented out for now. ([code/modules/mob/living/simple_animal/rogue/creacher/minotaur.dm](https://github.com/GeneralPantsuIsBadAtCoding/Azure-Peak/pull/996/files#diff-b38b58b374339dc1dfb3142eaaec05c3e329bbb50692bea8906a0d9a91f3e594))
- Reduced the chances of mobs idle sound playing slightly
- Altered the damage balance of the penultimate mobs slightly (minotaur, dragon) - this may need to be tweaked up further. Increased the reach of their weapons from 1 to 2
- Troll punches increased slightly.
- Trialling improved AI marauder (spearmen) orcs. I think this code is somewhat taxing.
- Dragon firebreath shifted over to spell but is unused at the moment. (chose action system as spells seemed to be a little bit of a mess. Can easily change over in the future if resolved)
- Some runtime errors fixed with carbon mobs
- Fixed some typos, magic numbers in code where I encountered them

Known Issues
- If you pause the game while debugging it, it may spew out a multitude of null errors from the AI system - a timer or something. I'm unsure as of yet what causes it as the message is not helpful.
- Shooting a mob from very far away, 1+ screens will generally not 'aggro' or activate it. 
- I don't know best practices for BYOND code so some may code not be as efficient where it can be. 
- Multiplayer performance: unknown, single player: within acceptable limits

This PR has gone far byond its scope.

bearrrrrrrr commented:
https://www.youtube.com/watch?v=z-JRdRXiNv4

Lutowski commented:
zepic

crix870 commented:
Mobs may need nerfs

Lutowski commented:
> Mobs may need nerfs

specifically for attack speed yes, last I checked there's no way to alter that for mobs. only npcs seem to have atk speed variables to edit

crix870 commented:
> > Mobs may need nerfs
> 
> specifically for attack speed yes, last I checked there's no way to alter that for mobs. only npcs seem to have atk speed variables to edit

You can give them custom intents and make them slower. I think all of the mobs (on live) are using 1s generic attacks

crix870 commented:
I've centralized all the general mob data (health, attack speed and movement speed) into defines so future-people can balance it to their heart's content.

Lutowski commented:
testmerging for a little bit, poor pr's waited long enough already

Lutowski commented:
fullmerging this, thank you
